File: Surrounded_Regions.cpp
/*130. Surrounded Regions  QuestionEditorial Solution  My Submissions
Total Accepted: 60534
Total Submissions: 362069
Difficulty: Medium
Given a 2D board containing 'X' and 'O' (the letter O), capture all regions surrounded by 'X'.

A region is captured by flipping all 'O's into 'X's in that surrounded region.

For example,
X X X X
X O O X
X X O X
X O X X
After running your function, the board should be:

X X X X
X X X X
X X X X
X O X X
Subscribe to see which companies asked this question

Show Tags
Show Similar Problems
*/


/*First, check the four border of the matrix. If there is a element is
'O', alter it and all its neighbor 'O' elements to '1'.
Then ,alter all the 'O' to 'X'
At last,alter all the '1' to 'O'*/

/////////////////////////////////////////////////////////////////////////////////////
//C++
class Solution {
public:
    void solve(vector<vector<char>>& board) {
       int i, j;
        int row = board.size();
        if(!row){
            return;
        }
        
        int col = board[0].size();
        for(i = 0; i < row; i++){
            check(board, i, 0, row, col);
            if(col>1){
                check(board, i, col-1, row, col);
            }
        }
        
        for(j = 1; j + 1 < col; j++){
            check(board, 0, j, row, col);
            if(row>1){
                check(board, row-1, j, row, col);
            }
        }
        
        for(i = 0; i < row; i++){
            for(j = 0; j < col; j++){
                if(board[i][j] == 'O'){
                    board[i][j] = 'X';
                }
            }
        }
        
        for(i = 0; i < row; i++){
            for(j = 0; j < col; j++){
                if(board[i][j] == '1'){
                    board[i][j] = 'O';
                }
            }
        }
    }
    
    void check(vector<vector<char> > &board, int i, int j, int row, int col){
        if(board[i][j]=='O'){
            board[i][j] = '1';
            if(i>1) check(board, i-1, j, row, col);
            if(j>1) check(board, i, j-1, row, col);
            if(i+1<row) check(board, i+1, j, row, col);
            if(j+1<col) check(board, i, j+1, row, col);
        }
    }
};







/////////////////////////////////////////////////////////////////////////////////////
//Java
public class Solution {
    public void solve(char[][] board) {
        int i, j;
        int row = board.length;
        if(row == 0){
            return;
        }
        
        int col = board[0].length;
        for(i = 0; i < row; i++){
            check(board, i, 0, row, col);
            if(col>1){
                check(board, i, col-1, row, col);
            }
        }
        
        for(j = 1; j + 1 < col; j++){
            check(board, 0, j, row, col);
            if(row>1){
                check(board, row-1, j, row, col);
            }
        }
        
        for(i = 0; i < row; i++){
            for(j = 0; j < col; j++){
                if(board[i][j] == 'O'){
                    board[i][j] = 'X';
                }
            }
        }
        
        for(i = 0; i < row; i++){
            for(j = 0; j < col; j++){
                if(board[i][j] == '1'){
                    board[i][j] = 'O';
                }
            }
        }
    }
    
    public void check(char[][] board, int i, int j, int row, int col){
        if(board[i][j] == 'O'){
            board[i][j] = '1';
            if(i>1) check(board, i-1, j, row, col);
            if(j>1) check(board, i, j-1, row, col);
            if(i+1<row) check(board, i+1, j, row, col);
            if(j+1<col) check(board, i, j+1, row, col);
        }
    }
}

File: Surrounded_Regions.swift
/*
Given an m x n matrix board containing 'X' and 'O', capture all regions surrounded by 'X'.

A region is captured by flipping all 'O's into 'X's in that surrounded region.

 

Example 1:


Input: board = [["X","X","X","X"],["X","O","O","X"],["X","X","O","X"],["X","O","X","X"]]
Output: [["X","X","X","X"],["X","X","X","X"],["X","X","X","X"],["X","O","X","X"]]
Explanation: Surrounded regions should not be on the border, which means that any 'O' on the border of the board are not flipped to 'X'. Any 'O' that is not on the border and it is not connected to an 'O' on the border will be flipped to 'X'. Two cells are connected if they are adjacent cells connected horizontally or vertically.
Example 2:

Input: board = [["X"]]
Output: [["X"]]
 

Constraints:

m == board.length
n == board[i].length
1 <= m, n <= 200
board[i][j] is 'X' or 'O'.
*/

/*
Solution 1:
BFS

search all connected O cells
mark if they should be flipped
if so, flip them

Time Complexity: O(mn)
Space Complexity: O(mn)
*/
class Solution {
    var dir = [0,-1,0,1,0]
    
    func solve(_ board: inout [[Character]]) {
        let m = board.count
        let n = board[0].count

        var visited = Array(repeating: Array(repeating: false, count: n),
                           count: m)
        
        for i in 0..<m {
            for j in 0..<n {
                if board[i][j] == "O", !visited[i][j] {
                    visited[i][j] = true
                    flipIfNeeded(i, j, m, n, &visited, &board)
                }
            }
        }
    }
    
    func flipIfNeeded(_ i: Int, _ j: Int, 
                     _ m: Int, _ n: Int,
                     _ visited: inout [[Bool]],
                     _ board: inout [[Character]]) {
        var connected = [(i, j)]
        
        var queue = [(i, j)]
        var notFlip = (i == 0 || j == 0 || i == m-1 || j == n-1)
        
        while !queue.isEmpty {
            let (r, c) = queue.removeFirst()
            
            for d in 0..<4 {
                let r1 = r+dir[d]
                let c1 = c+dir[d+1]
                
                if r1 >= 0, r1 < m, c1 >= 0, c1 < n, 
                !visited[r1][c1], board[r1][c1] == "O" {
                    visited[r1][c1] = true
                    
                    // find a connected cell which is not visited yet
                    connected.append((r1, c1))
                    queue.append((r1, c1))
                    
                    // update notFlip if needed
                    // print(r1, c1, r1 == m-1)
                    if !notFlip, (r1 == 0 || c1 == 0 || r1 == m-1 || c1 == n-1) { 
                        notFlip = true 
                    }
                }
            }
        }
        
        // print(i, j, notFlip, connected)
        // flip connected cell if needed
        if !notFlip {
            for (r, c) in connected {
                board[r][c] = "X"
            }
        }
    }
}